File: day22/typescript/day22.ts
import { readFile } from 'fs';
import { promisify } from 'util';

import * as R from 'ramda';

type Coord = [number, number];
type Grid = string[][];

const move = [
    ([x, y]: Coord) => [x, y - 1], // Up
    ([x, y]: Coord) => [x + 1, y], // Right
    ([x, y]: Coord) => [x, y + 1], // Down
    ([x, y]: Coord) => [x - 1, y], // Left
];

const travel = (grid: Grid, moves: number, next: (cell: string) => [number, string]) => {
    let infected = 0;
    let dir = 0;
    let [x, y] = [Math.floor(grid.length / 2), Math.floor(grid.length / 2)];
    for (let i = 0; i < moves; ++i) {
        if (grid[y] === undefined) {
            grid[y] = [];
        }
        if (grid[y][x] === undefined) {
            grid[y][x] = '.';
        }

        const [turn, set] = next(grid[y][x]);
        if (set === '#') {
            infected++;
        }

        grid[y][x] = set;
        dir = R.mathMod(dir + turn, 4);
        [x, y] = move[dir]([x, y]);
    }

    return infected;
};

const part1 = (grid: Grid) => travel(grid, 10000, cell => (
    cell === '#' ? [1, '.'] : [-1, '#']
));

const part2 = (grid: Grid) => travel(grid, 10000000, cell => {
    switch (cell) {
        case '#':
            return [1, 'F'];
        case 'W':
            return [0, '#'];
        case 'F':
            return [2, '.'];
        default:
            return [-1, 'W'];
    }
});

(async () => {
    const input = await promisify(readFile)('day22/input.txt', 'utf8');
    const grid = R.map(Array.from, R.split('\r\n', input));

    console.log(part1(grid.map(r => [...r])));
    console.log(part2(grid.map(r => [...r])));
})();

File: day22/haskell/day22.hs
{-# LANGUAGE LambdaCase #-}

import Data.Map (Map, findWithDefault, insert, fromList)

data Cell = I | C | W | F deriving (Eq, Show)
type Coord = (Int, Int)
type Grid = Map Coord Cell

parse :: String -> Grid
parse = fromList . concatMap f . zip [0..] . map (zip [0..]) . lines
    where f (y, xs)  = map (g y) xs
          g y (x, c) = ((x, y), if c == '#' then I else C)

move :: Int -> Coord -> Coord
move 0 (x, y) = (x, y - 1)
move 1 (x, y) = (x + 1, y)
move 2 (x, y) = (x, y + 1)
move _ (x, y) = (x - 1, y)

travel :: (Cell -> (Int, Cell)) -> Int -> Int -> Coord -> Grid -> Int -> Int
travel _ 0 _ _ _ i = i
travel f n d c g i
    | s == I    = next (i + 1)
    | otherwise = next i
    where (t, s) = f (findWithDefault C c g)
          dir    = ((d + t) `mod` 4)
          next   = travel f (n - 1) dir (move dir c) (insert c s g)

part1 :: Grid -> Int
part1 g = travel f 10000 0 (12, 12) g 0 
    where f = \case
                I -> (1, C)
                _ -> (-1, I)

part2 :: Grid -> Int
part2 g = travel f 10000000 0 (12, 12) g 0 
    where f = \case
                I -> (1, F)
                W -> (0, I)
                F -> (2, C)
                C -> (-1, W)

main :: IO ()
main = do
    grid <- parse <$> readFile "day22/input.txt"
    print . part1 $ grid
    print . part2 $ grid